File: src/test/java/destiny/utils/ZoneOffsetTest.java
/**
 * Created by smallufo on 2017-03-04.
 */
package destiny.utils;

import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;

import static org.junit.Assert.assertEquals;

public class ZoneOffsetTest {

  private Logger logger = LoggerFactory.getLogger(getClass());

  @Test
  public void testOf_8H() {
    ZoneOffset zf = ZoneOffset.of("+8");
    logger.info("zf = {} , id = {}", zf, zf.getId());
    assertEquals("+08:00", zf.getId());
    assertEquals(8 * 60 * 60, zf.getTotalSeconds());

    ZoneId zid = ZoneId.of("Asia/Taipei");
    zid.normalized();
  }

  /**
   * 8小時零1分
   */
  @Test
  public void testOf_8H1m() {
    ZoneOffset zf = ZoneOffset.of("+08:01");
    logger.info("zf = {} , id = {}", zf, zf.getId());
    assertEquals("+08:01", zf.getId());
    assertEquals(8 * 60 * 60 + 60, zf.getTotalSeconds());
  }

  /**
   * 8小時零1分2秒
   */
  @Test
  public void testOf_8H1m2s() {
    ZoneOffset zf = ZoneOffset.ofHoursMinutesSeconds(8, 1, 2);
    logger.info("zf = {} , id = {}", zf, zf.getId());
    assertEquals("+08:01:02", zf.getId());
    assertEquals(8 * 60 * 60 + 60 + 2, zf.getTotalSeconds());
  }

  /**
   * Asia/Taipei 於 1975 夏天，有實施 日光節約時間 , zf 應該為 "+09:00"
   */
  @Test
  public void testAsiaTaipei() {
    LocalDateTime lmt = LocalDateTime.of(1976, 6, 1, 0, 0);
    ZoneOffset zf = lmt.atZone(ZoneId.of("Asia/Taipei")).getOffset();
    logger.debug("zf = {} , id = {} , 秒數 = {}", zf, zf.getId(), zf.getTotalSeconds());
    assertEquals("+08:00", zf.getId());
    assertEquals(8 * 60 * 60, zf.getTotalSeconds());

    lmt = LocalDateTime.of(1975, 6, 1, 0, 0);
    zf = lmt.atZone(ZoneId.of("Asia/Taipei")).getOffset();
    logger.debug("zf = {} , id = {} , 秒數 = {}", zf, zf.getId(), zf.getTotalSeconds());
    assertEquals("+09:00", zf.getId());
    assertEquals(9 * 60 * 60, zf.getTotalSeconds());
  }
}

File: src/test/kotlin/destiny/tools/ZoneOffsetTest.kt
/**
 * Created by smallufo on 2017-03-04.
 */
package destiny.tools

import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import kotlin.test.Test
import kotlin.test.assertEquals

class ZoneOffsetTest {

  private val logger = KotlinLogging.logger {  }

  @Test
  fun testOf_8H() {
    val zf = ZoneOffset.of("+8")
    logger.info("zf = {} , id = {}", zf, zf.id)
    assertEquals("+08:00", zf.id)
    assertEquals((8 * 60 * 60).toLong(), zf.totalSeconds.toLong())

    val zid = ZoneId.of("Asia/Taipei")
    zid.normalized()
  }

  /**
   * 8小時零1分
   */
  @Test
  fun testOf_8H1m() {
    val zf = ZoneOffset.of("+08:01")
    logger.info("zf = {} , id = {}", zf, zf.id)
    assertEquals("+08:01", zf.id)
    assertEquals((8 * 60 * 60 + 60).toLong(), zf.totalSeconds.toLong())
  }

  /**
   * 8小時零1分2秒
   */
  @Test
  fun testOf_8H1m2s() {
    val zf = ZoneOffset.ofHoursMinutesSeconds(8, 1, 2)
    logger.info("zf = {} , id = {}", zf, zf.id)
    assertEquals("+08:01:02", zf.id)
    assertEquals((8 * 60 * 60 + 60 + 2).toLong(), zf.totalSeconds.toLong())
  }

  /**
   * Asia/Taipei 於 1975 夏天，有實施 日光節約時間 , zf 應該為 "+09:00"
   */
  @Test
  fun testAsiaTaipei() {
    var lmt = LocalDateTime.of(1976, 6, 1, 0, 0)
    var zf = lmt.atZone(ZoneId.of("Asia/Taipei")).offset
    logger.debug("zf = {} , id = {} , 秒數 = {}", zf, zf.id, zf.totalSeconds)
    assertEquals("+08:00", zf.id)
    assertEquals((8 * 60 * 60).toLong(), zf.totalSeconds.toLong())

    lmt = LocalDateTime.of(1975, 6, 1, 0, 0)
    zf = lmt.atZone(ZoneId.of("Asia/Taipei")).offset
    logger.debug("zf = {} , id = {} , 秒數 = {}", zf, zf.id, zf.totalSeconds)
    assertEquals("+09:00", zf.id)
    assertEquals((9 * 60 * 60).toLong(), zf.totalSeconds.toLong())
  }
}
